pkg/liqo-controller-manager/shadowendpointslice-controller/shadowendpointslice_controller.go

```diff
@@ -0,0 +1,244 @@
+		if errors.IsNotFound(err) {
+			klog.V(4).Infof("shadowendpointslice %s not found", nsName)
+			return ctrl.Result{}, nil
+		}
+		klog.Errorf("an error occurred while getting shadowendpointslice %s: %v", nsName, err)
+		return ctrl.Result{}, err
+	}
+
+	// Get ForeignCluster associated with the shadowendpointslice
+	clusterID, ok := shadowEps.Labels[forge.LiqoOriginClusterIDKey]
+	if !ok {
+		klog.Errorf("shadowendpointslice %s has no label with origin clusterID", nsName)
+		return ctrl.Result{}, fmt.Errorf("shadowendpointslice %s has no label with origin clusterID", nsName)
+	}
+	fc, err := foreigncluster.GetForeignClusterByID(ctx, r.Client, clusterID)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			klog.Error("foreigncluster associated with shadowendpointslice %s not found", nsName)
+			return ctrl.Result{}, err
+		}
+		klog.Errorf("an error occurred while getting foreigncluster %s: %v", clusterID, err)
+		return ctrl.Result{}, err
+	}
+
+	// Check network status of the foreigncluster
+	networkStatus := foreigncluster.IsNetworkingEstablished(fc)
+
+	// Based on the current network status to the foreign cluster, we update all endpoints' "Ready" condition
+	// Note: An endpoint is updated only if the shadowendpointslice endpoint has the condition "Ready" set
+	// to True or nil. i.e: if the foreigncluster sets the endpoint condition "Ready" to False, also the local
+	// endpoint condition is set to False regardless of the current network status
+	for i := range shadowEps.Spec.Template.Endpoints {
+		endpoint := &shadowEps.Spec.Template.Endpoints[i]
+		if endpoint.Conditions.Ready == nil || *endpoint.Conditions.Ready {
+			endpoint.Conditions.Ready = &networkStatus
+		}
+	}
+
+	// update shadowendpointslice labels to include the "managed-by"
+	shadowEps.SetLabels(labels.Merge(shadowEps.Labels, labels.Set{liqoconsts.ManagedByLabelKey: liqoconsts.ManagedByShadowEndpointSliceValue}))
+
+	// if any existing endpointslice is already been created from the shadowendpointslice...
+	var existingEps discoveryv1.EndpointSlice
+	if err := r.Get(ctx, nsName, &existingEps); err != nil && !errors.IsNotFound(err) {
+		klog.Errorf("unable to get endpointslice %s: %v", nsName, err)
+		return ctrl.Result{}, err
+	} else if err == nil {
+		// Update Labels and Annotations
+		klog.V(4).Infof("endpointslice %q found running, will update it", klog.KObj(&existingEps))
+
+		// Create Apply object for Existing Pod
+		epsApply := EndpointSliceApply(&shadowEps)
+
+		if err := r.Patch(ctx, &existingEps, clientutils.Patch(epsApply),
+			client.ForceOwnership, client.FieldOwner(ctrlFieldManager)); err != nil {
+			klog.Errorf("unable to update endpointslice %q: %v", klog.KObj(&existingEps), err)
+			return ctrl.Result{}, err
+		}
+
+		klog.Infof("updated endpointslice %q with success", klog.KObj(&existingEps))
+
+		return ctrl.Result{}, nil
+	}
+
+	// Else, create a new EndpointSlice
+	newEps := discoveryv1.EndpointSlice{
```

Comment:
It's probably better (for readability reasons) to create this resource at the beginning and then operate over it